### ISSUE TITLE
execute teardown logic only if element is removed

### ISSUE THREAD

**Issue by jelhan** [spencer516/ember-functional-modifiers#8]
First of all: Thanks a lot for this awesome addon! Instantly inspired me to write [`ember-style-modifier`](https://github.com/jelhan/ember-style-modifier). While doing so, I run into a limitation of current teardown functionality.

Is there any way to execute teardown logic only if element was removed from DOM but not if an argument changes? Actually I expected it to work like that when reading docs but noticed that teardown function is executed also on every change - and that this is by intention: https://github.com/spencer516/ember-functional-modifiers/blob/40be2487ec1699e4aae8b68bd7a26842f8836294/tests/integration/modifier-managers/functional-modifiers-test.js#L96-L115

I expect many use cases that require different logic to be execute if element is removed from DOM or if argument changes. My use case was the following: I needed to share state between different executions of the modifier. Especially it needs to know if it was executed before and if so, which arguments were used. I tried to implement a simple store using a `Map` to hold that information. To not end up with a memory leak the element needs to be removed from this `Map` if it's removed from DOM. I tried to use the teardown function therefor but as this was executed also on every change it doesn't work out.

**Comment by spencer516** [comment]
Ah that's awesome! I like the use case with `ember-style-modifiers`. 

I think this is a really good question — I went this route just to be consistent with how React was thinking about the `useEffect` hook: https://reactjs.org/docs/hooks-effect.html#explanation-why-effects-run-on-each-update 

I suppose one way we could handle this would be to pass an argument to the teardown method — something like "isRemoving", so you could write some defensive code to check whether it's tearing down: 

```js
export default makeFunctionalModifier(element => {
  // setup event
  return isRemoving => {
    if (isRemoving) {
      // teardown event
    }
  }
});
```

The problem here is that the closure that's created during the setup is the same one that'd be needed for the teardown (but not for the updates). For example, if we were using something ala RxJS (let's assume the second argument to the modifier is the `isInserting` flag):

```js
export default makeFunctionalModifier((element, isInserting, [action]) => {
  let observer;

  if (isInserting) {
    observer = makeRxJSThingy(element, action);    
  }

  return isRemoving => isRemoving && observer.unsubscribe();
});
```

This will work and teardown properly _provided the action never changes_. If the action changes, then `observer` will be undefined when we try to execute the teardown method. Blerg.

---------


If we wanted to make this work, then we might use something like a `WeakMap` instead to attach some state to the `element`. That might look something like this:

(An aside: using a `WeakMap` _should_ solve the memory leak concerns you were describing!)

```js
const OBSERVERS = new WeakMap();

export default makeFunctionalModifier((element, [action]) => {
  let observer;
  
  if (OBSERVER.has(element)) {
    observer = OBSERVER.get(element);
  } else {
    observer = makeRxJSThingy(element, action);
  }

  OBSERVER.set(element, observer);

  return isRemoving => isRemoving && OBSERVER.get(element).unsubscribe();
})
```

I have a few thoughts about an approach like this:

- Is this better than just using [ember-oo-modifiers](https://github.com/sukima/ember-oo-modifiers)? Not sure — maybe that provides a better API for when there's some state needed between updates. 
- Is there a way to make state a first-class thing with functional modifiers? (ala [`useState`](https://reactjs.org/docs/hooks-state.html)?) Maybe that would be nice (in conjunction with an `isRemoving` flag?):

```js
export default makeFunctionalModifier((element, [action]) => {
  const [observer] = useState(() => makeRxJSThingy(element, action));
  return isRemoving => isRemoving && observer.unsubscribe();
});
```

-------

Anyway — would love to hear your thoughts! Any other ways we might approach this? Do you have any use cases that I'm not thinking about?

**Comment by jelhan** [comment]
I had done a similar implementation to your second example using a `WeakMap()`. I've pushed it to GitHub now, so you could see real world use case: https://github.com/jelhan/ember-style-modifier/pull/6 Providing a `isRemoving` flag should be enough to make it working.

State handling is a complex concern IMO. I would tend to not support it out of the box but let the consumers experiment with different approaches. I'm a little bit afraid of too much breaking changes if `ember-functional-modifiers` getting to specific right now. Also it seems to be quite obvious how to handle the state as both of us were coming up with a similar solution.

Wasn't aware of `ember-oo-modifiers` at all. IMO that one feels more like the ember way of doing it as the lifecycle hooks are similar to components. Maybe I'm rewriting `ember-style-modifier` with `ember-oo-modifers` and have a look how it works out. But I think there is still a need for an `isRemoving` flag here - even if it's not about state management.

TL;DR: IExposing an `isRemoving` flag seems to be enough for now. We might consider adding something like `useState` later.

But I'm not sure about the naming yet. Is the teardown function called before or after the element has been removed from DOM? Reading the docs I thought it's executed afterwards: "*If the functionality you add in the modifier needs to be torn down when the element is removed, you can return a function for the teardown method.*" If it's executed after removing from DOM I would call the flag `isRemoved`. Otherwise the docs should be changed to *"will be removed"* or something similar.

**Comment by spencer516** [comment]
@jelhan — I tend to agree with your thoughts about managing state. (With this "functional"-ish approach to modifiers, state starts to get weird — as does functional programming in javascript generally).

> Is the teardown function called before or after the element has been removed from DOM? 

It corresponds with the [`destroyModifier`](https://emberjs.github.io/rfcs/0373-Element-Modifier-Managers.html#destroymodifier) hook. From the RFC: 

> [it] may or may not be called in the same tick as DOM removal

So — I'm not sure there's a guarantee there.

**Comment by spencer516** [comment]
(Also, opened a PR to add the flag — take a look and give it a try if you can!: https://github.com/spencer516/ember-functional-modifiers/pull/9 )
